### PR TITLE
Driver RTC for the board SAML21 by FWX

### PR DESCRIPTION
**Pull request by bapclenet**
FWX did a port of RTC from the SAMR21 to the SAML21. 

TODO: Need to be tested

### CODE REVIEW

**Comment by Troels51** [comment]
I can't really test this, because I don't have that hardware

**Comment by bapclenet** [comment]
@Troels51, I thought you did the port to SAML21. @kaspar030, could you try it? I don't have SAML21 board yet

**Comment by kaspar030** [comment]
@bapclenet Currently I don't have one either. @emmanuelsearch Didn't we buy one some time ago?

**Comment by OlegHahm** [comment]
@kaspar030, yes, I have two of them here. Will test.

**Comment by kaspar030** [comment]
AFAIK the SAM L21 has a shared RTT/RTC module, so this PR would collide with the existing RTT implementation.

Please see #2538 and #4565.

I just talked offline to @haukepetersen, he agrees that RTT should be preferred, as it can easily be used as base for RTC functionality in software (as in #4565).

The RTC interface is supposed to be used only if the underlying hardware module cannot be used in counter/RTT mode.

**Comment by bapclenet** [comment]
@kaspar030, @haukepetersen, can we configure the RTT driver to use the external Oscillator? XOSCK32 (as RTC does) This is important for accurate RTC

**Comment by kaspar030** [comment]
@bapclenet "can" as in "is it possible"? When I did the implementation, minimal power consumption was premium, so currently RTT is hard coded to ULP32K. We can change that, or better, make it configurable. Would you like to take a look at that?

**Comment by OlegHahm** [comment]
FYI: testing revealed that the RTC on my board here has a drift of about -23333 ppm.

**Comment by kaspar030** [comment]
> FYI: testing revealed that the RTC on my board here has a drift of about -23333 ppm.

Yeah, but that board only consumes <2nA when setting deep sleep mode. Changing the used OSC should vastly improve RTT accuracy.

**Comment by kaspar030** [comment]
@OlegHahm sorry, were you talking about this PR?

**Comment by OlegHahm** [comment]
Yes.

**Comment by kaspar030** [comment]
@OlegHahm Can you somehow test the accuracy of the RTT?

**Comment by OlegHahm** [comment]
If there doesn't exist an application for that already, not this week.

**Comment by bapclenet** [comment]
@OlegHahm, can you do the same test with the SAMR21? There are very similar.
@kaspar030, yes it was "is it possible"

Yes we could mix both but we need to configure it somewhere

**Comment by OlegHahm** [comment]
RTC on my SAMR21 runs without any notable drift.

**Comment by bapclenet** [comment]
I removed Ready for CI build since we haven't found an agreement

**Comment by bapclenet** [comment]
So there is a problem with this driver anyway. @OlegHahm, Which test have you run?
@Dialine, may you test on your board as you wrote the driver?

**Comment by OlegHahm** [comment]
Just manual testing using the `default` example and the `rtc gettime` command.

**Comment by Dialine** [comment]
Hi Oleg,

 I test this driver on SAML21-Xplained pro board, with test/periph_rtc.

 Attached, the trace of execution done here.

 Can you give me more information about the problem?

 Regards,
              FWX.



Le 22/01/2016 10:47, Oleg Hahm a écrit :
> Just manual testing using the |default| example and the |rtc gettime|
> command.
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RIOT-OS/RIOT/pull/4655#issuecomment-173863616>.
> 

main(): This is RIOT! (Version: 2016.03-devel-85-g369233-Dell64)

RIOT RTC low-level driver test
This test will display 'Alarm' in 10 seconds

Initializing the RTC driver
Setting clock to 2011-12-13 14:15:15
Clock set to 2011-12-13 14:15:15
Setting alarm to 2011-12-13 14:15:25
Alarm set to 2011-12-13 14:15:25
The alarm should trigger every 10 seconds for 4 times.
Alarm!
Alarm!
Alarm!
Alarm!

**Comment by OlegHahm** [comment]
Flash the [default example](https://github.com/RIOT-OS/RIOT/tree/master/examples/default), call `make term` and then call `rtc gettime` inside the RIOT shell, wait 10 minutes and call `rtc gettime` again. Now compare the two timestamps to the timestamps of pyterm. On my board I observed a significant drift.

**Comment by Dialine** [comment]
Ok Oleg,

 With default program, I type these commands :
 rtc init
 rtc poweron
 rtc settime 2016-01-22 14:38:00
 rtc gettime

 and I can see a difference of some seconds every minute.

 I am going to investigate about this leeway.

             FWX.

**Comment by bapclenet** [comment]
@Dialine, have you found the problem out?

**Comment by Dialine** [comment]
There is a a significant drift, and it is possible to correct it, by
adding or subtracting a value to the RTC prescaler. The problem is to
find the good value for this correction, depending of crystal frequency.

 Value to add in RTC->MODE2.FREQCORR.reg

**Comment by Dialine** [comment]
Hello,

 Now, the RTC driver is ok, with three different sources of oscillator :

 - external
 - low power internal
 - ultra low power internal.

 No drift observed with external oscillator (on saml21-xplained pro
board) or low power internal.

 Just change compilation directive at the top of the rtc.c file.

 Regards,

        FWX.

**Comment by OlegHahm** [comment]
Cannot test this week, because I'm not in the office. @kaspar030, @haukepetersen, @thomaseichinger, @emmanuelsearch or anyone else at the university, can you test?

File: cpu/saml21/periph/rtc.c [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+#include "periph_conf.h"
+#include "sched.h"
+#include "thread.h"
+
+/* guard file in case no RTC device was specified */
+#if RTC_NUMOF
+
+#define EXTERNAL_OSC32_SOURCE                    1
+#define INTERNAL_OSC32_SOURCE                    0
+#define ULTRA_LOW_POWER_INTERNAL_OSC_SOURCE      0
```
Comment (bapclenet) [inline cpu/saml21/periph/rtc.c]: This should be availabe in periph_conf.h with 
```
    /**
     * @name RTC configuration
     * @{
     */
    #define RTC_NUMOF           (1U)
    #define RTC_DEV             RTC->MODE2
    /** @} */
```

**Comment by bapclenet** [comment]
I don't have this board, could someone give a test and ack this PR?

**Comment by OlegHahm** [comment]
@bapclenet, can you please squash and prefix the commit message? I can test next Tuesday.

**Comment by bapclenet** [comment]
Squash: done

**Comment by OlegHahm** [comment]
Travis is complaining about trailing whitespaces:
```
cpu/saml21/periph/rtc.c:149: trailing whitespace.
+    
cpu/saml21/periph/rtc.c:149: trailing whitespace.
+    
```

**Comment by bapclenet** [comment]
Removed and squashed

**Comment by bapclenet** [comment]
Which comments are you talking about?

**Comment by haukepetersen** [comment]
I don't like the clock initialization code inside the driver, I think it belongs into the generic cpu code. But for now I can live with that.

Also there are some c++ `//` style comments left :-)

**Comment by bapclenet** [comment]
@haukepetersen You mean the #ifdef?
Feel free to create a PR if you want to change anything. I will remove c++ comment

**Comment by bapclenet** [comment]
c++ comment removed

**Comment by bapclenet** [comment]
ping @haukepetersen 
Let's merge?

**Comment by haukepetersen** [comment]
> @haukepetersen You mean the #ifdef?

actually i mean the whole initialization of the `OSC32KCTRL`, but as said, I won't object merging this for now. So go if you can convince Travis :-)

**Comment by OlegHahm** [comment]
Needs a rebase to make Travis happy.

**Comment by bapclenet** [comment]
Rebased, wait for Travis

**Comment by Dialine** [comment]
Merci Baptiste.

     FWX.

**Comment by bapclenet** [comment]
@hauke, I let you merge (I don't think I have the right to merge if I'm the owner of the PR, do I?)

**Comment by gebart** [comment]
If you have an ACK from someone else and Travis is fine then you can merge it regardless if you are the creator of the PR or not. However, we're missing a formal ACK from someone right now... (I don't have the hardware)

**Comment by OlegHahm** [comment]
I'm fine to ACK.

**Pull request merged**